src/network/propagate.rs

```diff
@@ -122,25 +125,56 @@ fn ensure_propagation_peer<'a>(
 impl PropagationMap {
     pub fn new() -> Self {
         PropagationMap {
-            lock: RwLock::new(HashMap::new()),
+            mutex: Mutex::new(HashMap::new()),
         }
     }
 
     pub fn subscribe_to_blocks(&self, id: p2p::NodeId) -> Subscription<Header> {
-        let mut map = self.lock.write().unwrap();
+        let mut map = self.mutex.lock().unwrap();
         let handles = ensure_propagation_peer(&mut map, id);
         handles.blocks.subscribe()
     }
 
     pub fn subscribe_to_messages(&self, id: p2p::NodeId) -> Subscription<Message> {
-        let mut map = self.lock.write().unwrap();
+        let mut map = self.mutex.lock().unwrap();
         let handles = ensure_propagation_peer(&mut map, id);
         handles.messages.subscribe()
     }
 
     pub fn subscribe_to_gossip(&self, id: p2p::NodeId) -> Subscription<Gossip<p2p::Node>> {
-        let mut map = self.lock.write().unwrap();
+        let mut map = self.mutex.lock().unwrap();
         let handles = ensure_propagation_peer(&mut map, id);
         handles.gossip.subscribe()
     }
+
+    fn propagate_with<F>(&self, ids: &[p2p::NodeId], f: F)
+    where
+        F: Fn(&mut PeerHandles) -> Result<(), PropagateError>,
+    {
+        let mut map = self.mutex.lock().unwrap();
+        for id in ids {
+            if let hash_map::Entry::Occupied(mut entry) = map.entry(id.clone()) {
+                f(entry.get_mut()).unwrap_or_else(|e| {
+                    info!("propagation stream error: {:?}", e);
```

Comment:
a `warn` would make more sense, no?

Reply:
It might happen for benign reasons, like loss of network connectivity or the peer being slow. The mitigation is to simply unsubscribe the peer.

Reply:
the keyword error in the log is confusing.